File: android/app/src/main/kotlin/ca/lucianlabs/tgsocial/ui/screens/SettingsScreen.kt
package ca.lucianlabs.tgsocial.ui.screens

import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import ca.lucianlabs.housepour.HPAvatar
import ca.lucianlabs.housepour.HPBody
import ca.lucianlabs.housepour.HPButton
import ca.lucianlabs.housepour.HPButtonSize
import ca.lucianlabs.housepour.HPButtonStyle
import ca.lucianlabs.housepour.HPCard
import ca.lucianlabs.housepour.HPListItem
import ca.lucianlabs.housepour.HPMonoSmall
import ca.lucianlabs.housepour.HPMuted
import ca.lucianlabs.housepour.HPSectionMark
import ca.lucianlabs.housepour.HPSmall
import ca.lucianlabs.housepour.HPText
import ca.lucianlabs.housepour.HPTokens
import ca.lucianlabs.tgsocial.demo.DemoCopy
import ca.lucianlabs.tgsocial.model.MyNode
import ca.lucianlabs.tgsocial.protocol.HiddenItem
import ca.lucianlabs.tgsocial.protocol.ReportEmail
import ca.lucianlabs.tgsocial.protocol.SafetyLists
import ca.lucianlabs.tgsocial.protocol.Username
import ca.lucianlabs.tgsocial.ui.AppViewModel
import ca.lucianlabs.tgsocial.ui.Screen
import ca.lucianlabs.tgsocial.ui.Sheet
import ca.lucianlabs.tgsocial.ui.columnItem
import ca.lucianlabs.tgsocial.ui.components.openMail
import ca.lucianlabs.tgsocial.ui.components.rememberTdImage

/**
 * PRODUCT §2.20 — Settings: the safety lists, the contact card, and the two destructive actions in that
 * order. Sign Out moved here from You (§2.8) so `Delete My Node` could sit under it rather than a mis-tap
 * away from `View as others see it` — the reversible destructive action first, the irreversible one second.
 */
fun LazyListScope.SettingsItems(vm: AppViewModel, safety: SafetyLists, node: MyNode?, inDemo: Boolean = false) {
    item(key = "settings-safety-mark") {
        Box(Modifier.columnItem().padding(bottom = HPTokens.Space.rowGap)) { HPSectionMark("Safety") }
    }
    item(key = "settings-safety-note") {
        Box(Modifier.columnItem().padding(bottom = HPTokens.Space.cardGap)) {
            HPMuted("Blocked and reported content is hidden everywhere in the app. The filter is always on; there is no switch. These lists live on this device only and nobody else can read them.")
        }
    }

    item(key = "settings-blocked-mark") {
        Box(Modifier.columnItem().padding(bottom = HPTokens.Space.rowGap)) { HPSectionMark("Blocked", safety.blocked.size) }
    }
    item(key = "settings-blocked") {
        ListCard(safety.blocked.isEmpty(), "You haven't blocked anyone.") {
            val cards by vm.cards.collectAsStateWithLifecycle()
            safety.blocked.forEachIndexed { i, username ->
                val snapshot = cards[Username.key(username)]
                val image = rememberTdImage(snapshot?.photo, HPTokens.Space.avatarRow)
                HPListItem(
                    modifier = Modifier.rowTap("Open @$username") { vm.push(Screen.Profile(username)) },
                    isLast = i == safety.blocked.lastIndex,
                    trailing = { HPButton("Unblock", { vm.unblock(username) }, style = HPButtonStyle.GHOST, size = HPButtonSize.SMALL, contentDescription = "Unblock @$username") },
                ) {
                    HPAvatar(image, HPTokens.Space.avatarRow, snapshot?.initial ?: "·", contentDescription = null)
                    Column(Modifier.weight(1f)) {
                        HPBody(snapshot?.displayName ?: "@$username", strong = true, maxLines = 1)
                        HPMonoSmall("@$username", maxLines = 1)
                    }
                }
            }
        }
    }

    item(key = "settings-muted-mark") {
        Box(Modifier.columnItem().padding(bottom = HPTokens.Space.rowGap)) { HPSectionMark("Muted", safety.mutedFeeds.size) }
    }
    item(key = "settings-muted") {
        ListCard(safety.mutedFeeds.isEmpty(), "No muted feeds.") {
            safety.mutedFeeds.forEachIndexed { i, username ->
                val title = vm.cachedFeedSource(username)?.title ?: "@$username"
                HPListItem(
                    modifier = Modifier.rowTap("Open @$username") { vm.push(Screen.FeedChannel(username)) },
                    isLast = i == safety.mutedFeeds.lastIndex,
                    trailing = { HPButton("Unmute", { vm.unmuteFeed(username, title) }, style = HPButtonStyle.GHOST, size = HPButtonSize.SMALL, contentDescription = "Unmute $title") },
                ) {
                    Column(Modifier.weight(1f)) {
                        HPBody(title, maxLines = 1)
                        HPMonoSmall("@$username", maxLines = 1)
                    }
                }
            }
        }
    }

    item(key = "settings-hidden-mark") {
        Box(Modifier.columnItem().padding(bottom = HPTokens.Space.rowGap)) { HPSectionMark("Hidden", safety.hidden.size) }
    }
    item(key = "settings-hidden") {
        ListCard(safety.hidden.isEmpty(), "Nothing hidden.") {
            safety.hidden.forEachIndexed { i, item -> HiddenRow(vm, item, isLast = i == safety.hidden.lastIndex) }
        }
    }

    item(key = "settings-contact-mark") {
        Box(Modifier.columnItem().padding(bottom = HPTokens.Space.rowGap)) { HPSectionMark("Contact") }
    }
    item(key = "settings-contact") {
        val context = LocalContext.current
        Box(Modifier.columnItem()) {
            HPCard {
                Box(
                    modifier = Modifier
                        .defaultMinSize(minHeight = HPTokens.Space.touchMin)
                        .clickable(interactionSource = remember { MutableInteractionSource() }, indication = null, role = Role.Button) {
                            openMail(context, ReportEmail.ADDRESS)
                        }
                        .semantics { contentDescription = "Write to ${ReportEmail.ADDRESS}" },
                    contentAlignment = Alignment.CenterStart,
                ) {
                    HPText(ReportEmail.ADDRESS, HPTokens.Type.body, HPTokens.Colors.accent, maxLines = 1)
                }
                Spacer(Modifier.height(HPTokens.Space.rowGap))
                // The honest part: a client with no server cannot delete someone else's channel, so this says
                // what it can do instead of implying a takedown it cannot perform (§2.19).
                HPMuted("Reports are read by a person within 24 hours. Content that breaks the rules is reported to Telegram, the only party that can remove it from the network. Your copy is hidden on your device the moment you report it, whether or not anyone else acts.")
            }
        }
    }

    // PRODUCT §2.22.3 — `Sign Out` is not in the demo at all: `( Leave Demo )` (neutral) sits where it sits in
    // a real session, above `( Delete My Node )` (danger). Leaving is not destructive, so it is not danger.
    item(key = "settings-signout") {
        Box(Modifier.columnItem()) {
            if (inDemo) HPButton(DemoCopy.LEAVE, { vm.leaveDemo() }, style = HPButtonStyle.NEUTRAL)
            else HPButton("Sign Out", { vm.openSheet(Sheet.SignOut) }, style = HPButtonStyle.DANGER)
        }
    }
    if (node != null) {
        item(key = "settings-delete-node") {
            Box(Modifier.columnItem().padding(top = HPTokens.Space.rowGap)) {
                HPButton("Delete My Node", { vm.openSheet(Sheet.DeleteNode) }, style = HPButtonStyle.DANGER)
            }
        }
    }
}

/**
 * A hidden row names its channel and message id and **never** the content: showing a preview of the thing
 * someone reported would undo the report (§2.20). The reason is the one they picked, kept verbatim.
 */
@Composable
private fun HiddenRow(vm: AppViewModel, item: HiddenItem, isLast: Boolean) {
    val channel = item.key.substringBefore('/')
    val messageId = item.key.substringAfter('/')
    val title = vm.cachedFeedSource(channel)?.title ?: "@$channel"
    HPListItem(
        isLast = isLast,
        trailing = { HPButton("Unhide", { vm.unhide(item.key) }, style = HPButtonStyle.GHOST, size = HPButtonSize.SMALL, contentDescription = "Unhide $title $messageId") },
    ) {
        Column(Modifier.weight(1f)) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(0.dp)) {
                HPBody(title, Modifier.weight(1f, fill = false), maxLines = 1)
                HPMonoSmall(" · $messageId", maxLines = 1)
            }
            // `at` is ISO 8601 UTC (PROTOCOL §7.1); the row shows the day, which is what "reported when" means.
            HPSmall("${item.reason} · reported ${item.at.take(10)}", maxLines = 1)
        }
    }
}

/** A list card, or its empty line — the shape every section in §2.20 shares. */
@Composable
private fun ListCard(empty: Boolean, emptyText: String, rows: @Composable () -> Unit) {
    Box(Modifier.columnItem()) {
        HPCard(padding = PaddingValues(horizontal = HPTokens.Space.cardPad, vertical = 0.dp)) {
            if (empty) {
                Spacer(Modifier.height(HPTokens.Space.rowPad))
                HPMuted(emptyText)
                Spacer(Modifier.height(HPTokens.Space.rowPad))
            } else {
                rows()
            }
        }
    }
}

@Composable
private fun Modifier.rowTap(label: String, onClick: () -> Unit): Modifier =
    this
        .clickable(interactionSource = remember { MutableInteractionSource() }, indication = null, role = Role.Button, onClick = onClick)
        .semantics { contentDescription = label }

File: ios/Sources/Screens/SettingsScreen.swift
// Screens — Settings (PRODUCT.md §2.20) and Delete my node (§2.21).
//
// Everything the safety lists (PROTOCOL §7.1) hold, each row with its own undo, plus the contact
// card (§2.19) and the two destructive actions. Sign Out lives here rather than on You so the
// irreversible action can sit directly below the reversible one instead of a mis-tap away from
// `View as others see it`.

import SwiftUI

struct SettingsScreen: View {
    @Environment(AppModel.self) private var model

    var body: some View {
        Screen(back: true) {
            HPSectionMark("Safety")
            HPMuted("Blocked and reported content is hidden everywhere in the app. The filter is always on; there is no switch. These lists live on this device only and nobody else can read them.")
                .padding(.bottom, HPTokens.Space.cardGap)

            blocked
            muted
            hidden
            // PRODUCT §2.33: between HIDDEN and CONTACT, present only when there is a private layer.
            if !model.isDemo { PrivateSettingsSection() }
            contact

            // §2.22.3: `Sign Out` is not in the demo at all — there is no session to leave.
            // `( Leave Demo )` is neutral and sits exactly where it sat, above the danger button.
            if model.isDemo {
                HPButton(DemoCopy.leaveButton, style: .neutral) { model.leaveDemo() }
                    .padding(.top, HPTokens.Space.rowGap)
            } else {
                HPButton("Sign Out", style: .danger) { model.modal = .signOut }
                    .padding(.top, HPTokens.Space.rowGap)
            }
            if model.myNode != nil {
                HPButton("Delete My Node", style: .danger) { model.modal = .deleteNode }
                    .padding(.top, HPTokens.Space.rowGap)
            }
        }
    }

    // MARK: Blocked

    @ViewBuilder private var blocked: some View {
        let list = model.moderation.lists.blocked
        HPSectionMark("Blocked", count: list.count)
        if list.isEmpty {
            HPCard { HPMuted("You haven't blocked anyone.") }
        } else {
            HPListCard {
                ForEach(Array(list.enumerated()), id: \.element) { i, username in
                    let node = model.node(username)
                    HPListItem(isLast: i == list.count - 1) {
                        Button { model.path.append(.profile(username: username)) } label: {
                            HStack(alignment: .center, spacing: HPTokens.Space.rowGap) {
                                NodeAvatar(photo: node?.photo, size: HPTokens.Space.avatarRow,
                                           initial: String((node?.displayName ?? username).prefix(1)))
                                VStack(alignment: .leading, spacing: 0) {
                                    HPBody(node?.displayName ?? "@" + username, strong: true).lineLimit(1)
                                    HPMonoSmall("@" + username).lineLimit(1)
                                }
                            }
                            .frame(minHeight: HPTokens.Space.touchMin)
                            .contentShape(Rectangle())
                        }
                        .buttonStyle(.plain)
                        .accessibilityLabel("Open \(node?.displayName ?? username)")
                    } trailing: {
                        HPButton("Unblock", style: .ghost, size: .small) { model.unblock(username) }
                    }
                }
            }
        }
    }

    // MARK: Muted

    @ViewBuilder private var muted: some View {
        let list = model.moderation.lists.mutedFeeds
        HPSectionMark("Muted", count: list.count)
        if list.isEmpty {
            HPCard { HPMuted("No muted feeds.") }
        } else {
            HPListCard {
                ForEach(Array(list.enumerated()), id: \.element) { i, key in
                    // PRODUCT §2.32: a private channel shows its title with a `Private` pill in
                    // place of a username; its key is `c/<id>` (PROTOCOL §7.2).
                    let privateId = PrivateLink.supergroupId(fromSourceKey: key)
                    let privateInfo = privateId.flatMap { model.privateChatId(supergroupId: $0) }.flatMap { model.privateChannel(chatId: $0) }
                    let feed = privateId == nil ? model.feedInfo(key) : privateInfo
                    let title = feed?.title ?? (privateId == nil ? "@" + key : "Private channel")
                    HPListItem(isLast: i == list.count - 1) {
                        Button { model.openFeed(sourceKey: key) } label: {
                            VStack(alignment: .leading, spacing: 0) {
                                HStack(spacing: HPTokens.Space.rowGap) {
                                    HPBody(title).lineLimit(1)
                                    if privateId != nil { HPPill("Private", tone: .neutral) }
                                }
                                if privateId == nil { HPMonoSmall("@" + key).lineLimit(1) }
                            }
                            .frame(minHeight: HPTokens.Space.touchMin, alignment: .leading)
                            .contentShape(Rectangle())
                        }
                        .buttonStyle(.plain)
                        .accessibilityLabel("Open feed \(title)")
                    } trailing: {
                        HPButton("Unmute", style: .ghost, size: .small) { model.unmute(sourceKey: key, title: title) }
                    }
                }
            }
        }
    }

    // MARK: Hidden

    @ViewBuilder private var hidden: some View {
        let list = model.moderation.lists.hidden
        HPSectionMark("Hidden", count: list.count)
        if list.isEmpty {
            HPCard { HPMuted("Nothing hidden.") }
        } else {
            HPListCard {
                ForEach(Array(list.enumerated()), id: \.element.key) { i, item in
                    HPListItem(isLast: i == list.count - 1) {
                        VStack(alignment: .leading, spacing: 0) {
                            // §2.20: the row names its channel and message id, never the content —
                            // showing a preview of the thing someone reported would undo the report.
                            HPBody(Self.hiddenTitle(item, model: model)).lineLimit(1)
                            HPMonoSmall("\(item.reason) \u{00B7} reported \(Moderation.reportedDate(item.at))")
                                .lineLimit(1)
                        }
                        .frame(minHeight: HPTokens.Space.touchMin, alignment: .leading)
                    } trailing: {
                        HPButton("Unhide", style: .ghost, size: .small) { model.unhide(item) }
                    }
                }
            }
        }
    }

    /// `WaveLoop devlog · 144` — the channel's title where it is known, its username where it is not.
    /// A private post's key is `c/<id>/<n>` (PROTOCOL §7.2); its row reads the channel title with
    /// the key column `c/<id> · <n>` (PRODUCT §2.32), never a username it does not have.
    static func hiddenTitle(_ item: HiddenItem, model: AppModel) -> String {
        if let (sourceKey, n) = Self.privateParts(item.key) {
            let title = PrivateLink.supergroupId(fromSourceKey: sourceKey)
                .flatMap { model.privateChatId(supergroupId: $0) }
                .flatMap { model.privateChannel(chatId: $0)?.title } ?? "Private channel"
            return "\(title) \u{00B7} \(sourceKey) \u{00B7} \(n)"
        }
        let parts = item.key.split(separator: "/", maxSplits: 1, omittingEmptySubsequences: false)
        let channel = String(parts.first ?? "")
        let id = parts.count > 1 ? String(parts[1]) : ""
        let title = model.feedInfo(channel)?.title ?? "@" + channel
        return id.isEmpty ? title : "\(title) \u{00B7} \(id)"
    }

    /// `c/<id>/<n>` → (`c/<id>`, `<n>`); nil for a username key.
    static func privateParts(_ key: String) -> (String, String)? {
        let parts = key.split(separator: "/", omittingEmptySubsequences: false)
        guard parts.count == 3, parts[0] == "c", PrivateLink.supergroupId(fromSourceKey: "c/" + parts[1]) != nil else { return nil }
        return ("c/" + parts[1], String(parts[2]))
    }

    // MARK: Contact (§2.19)

    @ViewBuilder private var contact: some View {
        HPSectionMark("Contact")
        HPCard {
            Button { model.contactByMail() } label: {
                Text(Moderation.contactAddress)
                    .hpStyle(HPType.body, color: HPTokens.Colors.accent)
                    .underline()
                    .frame(minHeight: HPTokens.Space.touchMin, alignment: .leading)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Write to \(Moderation.contactAddress)")
            HPMuted("Reports are read by a person within 24 hours. Content that breaks the rules is reported to Telegram, the only party that can remove it from the network. Your copy is hidden on your device the moment you report it, whether or not anyone else acts.")
                .padding(.top, HPTokens.Space.rowGap)
        }
    }
}

/// Delete my node (PRODUCT §2.21): the type-the-username confirm, the run, and the four ways it can
/// end short of success. A tap-to-confirm would not be proportional — this destroys two public
/// channels and releases their names.
struct DeleteNodeModal: View {
    @Environment(AppModel.self) private var model
    @State private var typed = ""
    @State private var running = false
    @State private var failure: DeleteNodeResult?

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            if let failure {
                failureView(failure)
            } else {
                confirm
            }
        }
    }

    private var username: String { model.myNode?.username ?? "" }
    private var repliesUsername: String { model.myCard?.replies ?? model.suggestedRepliesUsername }

    // MARK: The confirm

    /// PRODUCT §2.21's paragraph, grown per §2.33 when a private node exists: the private clause
    /// is derived, and the sentence about members is present only then.
    private var paragraph: String {
        if let clause = model.deleteNodePrivateClause {
            return "This deletes the channel @\(username), your comments channel @\(repliesUsername), \(clause) from Telegram. The public card other people read disappears, every post and comment in those channels goes with it, every member of your private channels loses them at once, and the public names are released for anyone to take. This cannot be undone."
        }
        return "This deletes the channel @\(username) and your comments channel @\(repliesUsername) from Telegram. The public card other people read disappears, every post and comment in those two channels goes with it, and the names are released for anyone to take. This cannot be undone."
    }

    @ViewBuilder private var confirm: some View {
        HPSectionMark("Delete my node")
        HPH2("Delete my node.")
        HPMuted(paragraph)
            .padding(.top, HPTokens.Space.rowGap)
        HPMuted("Your feed channels are not touched.")
            .padding(.top, HPTokens.Space.rowGap)
            .padding(.bottom, HPTokens.Space.cardPad)
        HPTextField("Type @\(username) to confirm", text: $typed, placeholder: "@\(username)", kind: .mono)
        HPButtonRow {
            HPButton(running ? "Deleting\u{2026}" : "Delete My Node", style: .danger,
                     enabled: !running && Moderation.confirmsDelete(typed, username: username)) { run() }
        } b: {
            HPButton("Cancel", style: .ghost, enabled: !running) { model.modal = nil }
        }
    }

    private func run() {
        guard !running else { return }
        running = true
        // §2.21: while the delete runs the modal cannot be dismissed — a scrim tap mid-delete would
        // leave the run with nowhere to report the outcome.
        model.modalLocked = true
        Task {
            let result = await model.deleteMyNode()
            model.modalLocked = false
            running = false
            switch result {
            // Both handled by the model: it wipes local state and toasts, or says `You're offline.`
            case .deleted, .offline: break
            default: failure = result
            }
        }
    }

    // MARK: The ways it ends short

    @ViewBuilder private func failureView(_ result: DeleteNodeResult) -> some View {
        HPSectionMark("Delete my node")
        switch result {
        case .notOwner(let name):
            HPMuted(Self.message(for: result) ?? "")
                .padding(.bottom, HPTokens.Space.cardPad)
            HPButton("Open in Telegram", style: .neutral) {
                model.modal = nil
                model.openInTelegram(DeepLink.chat(username: name))
            }
            HPButton("Close", style: .ghost) { model.modal = nil }
                .padding(.top, HPTokens.Space.rowGap)
        case .commentsFailed, .nodeFailed, .privateFailed, .commentsFailedAfterPrivate, .nodeFailedAfterPrivate:
            retry(Self.message(for: result) ?? "")
        case .deleted, .offline:
            EmptyView()
        }
    }

    /// The sentence each short ending shows (PRODUCT §2.21, §2.33), verbatim. Every one names
    /// what is still there, and only the two that stop before anything went say
    /// `Nothing was deleted.` — once the private channels have gone (PROTOCOL §11.4.12 puts them
    /// first, and `deleteChat` is for every member at once) no later refusal may say it.
    static func message(for result: DeleteNodeResult) -> String? {
        switch result {
        case .notOwner(let name):
            return "Telegram won't let you delete @\(name) \u{2014} only the channel's owner can. Open it in Telegram to see who owns it."
        case .commentsFailed(let name, let error):
            return "Couldn't delete @\(name) \u{2014} Telegram said: \(error). Nothing was deleted."
        case .nodeFailed(let name, let error):
            return "Your comments channel is gone. @\(name) is still there \u{2014} Telegram said: \(error)."
        // PRODUCT §2.33: the four private outcomes, one per point the run can stop at.
        case .privateFailed(let title, let error):
            return "Couldn't delete \(title) \u{2014} Telegram said: \(error). Nothing was deleted."
        case .commentsFailedAfterPrivate(let name, let replies, let error):
            return "Your private channels are gone. @\(replies) and @\(name) are still there \u{2014} Telegram said: \(error)."
        case .nodeFailedAfterPrivate(let name, let commentsWent, let error):
            let gone = commentsWent ? "Your private channels and your comments channel are gone." : "Your private channels are gone."
            return "\(gone) @\(name) is still there \u{2014} Telegram said: \(error)."
        case .deleted, .offline:
            return nil
        }
    }

    @ViewBuilder private func retry(_ message: String) -> some View {
        HPMuted(message)
            .padding(.bottom, HPTokens.Space.cardPad)
        HPButtonRow {
            HPButton(running ? "Deleting\u{2026}" : "Try Again", style: .danger, enabled: !running) {
                failure = nil
                run()
            }
        } b: {
            HPButton("Close", style: .ghost, enabled: !running) { model.modal = nil }
        }
    }
}
